*** upstream/Makefile~	2014-10-28 14:21:23.000000000 -0500
--- upstream/Makefile	2014-10-28 14:21:47.669735146 -0500
***************
*** 239,249 ****
  configure_binutils = --target=mips-unknown-elf
  
  ## newlib ##############################################################################
  
! version_newlib = 1.11.0
  url_newlib = ftp://sources.redhat.com/pub/newlib/newlib-$(version_newlib).tar.gz
! patches_newlib = newlib-mips.patch newlib-tzset.patch newlib-malloc.patch newlib-nomemcpy.patch newlib-unix.patch newlib-unistd.patch newlib-nestedvm-define.patch
  configure_newlib = --enable-multilib --target=mips-unknown-elf
  
  tasks/build_newlib: tasks/build_gcc
  
--- 239,249 ----
  configure_binutils = --target=mips-unknown-elf
  
  ## newlib ##############################################################################
  
! version_newlib = 1.20.0
  url_newlib = ftp://sources.redhat.com/pub/newlib/newlib-$(version_newlib).tar.gz
! patches_newlib = newlib-mips.patch newlib-tzset.patch newlib-malloc.patch newlib-nomemcpy.patch newlib-unix.patch newlib-unistd.patch newlib-nestedvm-define.patch newlib-sdata.patch newlib-new.patch
  configure_newlib = --enable-multilib --target=mips-unknown-elf
  
  tasks/build_newlib: tasks/build_gcc
  
